### PR TITLE
feat: Blake2s round precompile

### PR DESCRIPTION
**Pull request by storojs72**
This PR implements gadget and syscall for Blake2s round function. 

The integrated code that allows using this syscall in external sphinx programs is located in [zkvm](https://github.com/argumentcomputer/RustCrypto-hashes/tree/zkvm) branch of RustCrypto.

The following is a comparison of single Blake2s hashing inside Sphinx program without / with precompile:
```
summary: cycles=5989, e2e=2.196652208, khz=2726.42, proofSize=1789916 (without precompile)
summary: cycles=4401, e2e=2.104103333, khz=2091.63, proofSize=3188917 (with precompile)
```

**TODO**: 
- Explore if it makes sense to extend the precompile to round_x10 or any other possible imprevements from performance perspective.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -78,7 +78,7 @@ getrandom = "=0.2.14" # 0.2.15 depends on yanked libc 0.2.154
 hashbrown = { version = "0.14.5", features = ["serde"] }
 hex = "0.4.3"
 home = "0.5.9"
-hybrid-array = "0.2.0-rc"
+hybrid-array = "0.2.0-rc.9"
```
Comment (huitseeker) [inline Cargo.toml:81]: the resolution to 0.2.0-rc.9 should be automatic without this change

File: core/src/utils/prove.rs [hunk 2]
```diff
@@ -356,11 +356,11 @@ where
     OpeningProof<SC>: Send + Sync,
     ShardMainData<SC>: Serialize + DeserializeOwned,
 {
-    #[cfg(debug_assertions)]
-    {
-        let record_clone = record.clone();
-        machine.debug_constraints(pk, record_clone);
-    }
+    //#[cfg(debug_assertions)]
```
Comment (huitseeker) [inline core/src/utils/prove.rs:359]: Nit: remove before merge.
FYI, we have a cargo-profile called dev-ci, which will be as efficient to run as `--release`, but has debug assertions!

**Review by huitseeker — COMMENTED** [review]
This looks pretty good at first glance! I'll come back to this.

File: core/src/syscall/precompiles/blake2s/air.rs [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+                local.xor[i + 12].value,
+                R_4 as usize,
+                local.rotate_right[i + 12],
+                local.shard,
+                &local.channel,
+                local.is_real,
+            );
+        }
+
+        self.constrain_shuffled_indices(builder, &local.shuffled_indices, local.is_real);
```
Comment (wwared) [inline core/src/syscall/precompiles/blake2s/air.rs:194]: It seems like `local.shuffled_indices` is only used here, and this is just enforcing that the first 4 elements are 0 while the rest are 1.

I believe you can remove the `shuffled_indices` column and its related constraints without any impact. The shuffling is still properly constrained by the constant values of the `vX_shuffle_lookup` vectors offsetting the memory reads.

File: zkvm/precompiles/src/lib.rs [hunk 4]
```diff
@@ -47,4 +47,13 @@ extern "C" {
...
     pub fn syscall_bls12381_g2_double(p: *mut u32);
+    pub fn syscall_blake2s_xor_rotate_16(w: *mut u32);
+
+    pub fn syscall_blake2s_add_2(left: *mut u32, right: *const u32);
+    pub fn syscall_blake2s_add_3(left: *mut u32, right: *const u32);
+    pub fn syscall_blake2s_xor_rotate_right_16(left: *mut u32, right: *const u32);
+    pub fn syscall_blake2s_xor_rotate_right_12(left: *mut u32, right: *const u32);
+    pub fn syscall_blake2s_xor_rotate_right_8(left: *mut u32, right: *const u32);
+    pub fn syscall_blake2s_xor_rotate_right_7(left: *mut u32, right: *const u32);
+    pub fn syscall_blake2s_quarter_round(left: *mut u32, right: *const u32);
```
Comment (wwared) [inline zkvm/precompiles/src/lib.rs:58]: I believe these can be removed since they are no longer being used/defined

File: core/src/stark/air.rs [hunk 5]
```diff
@@ -181,6 +184,8 @@ impl<F: PrimeField32> RiscvAir<F> {
         chips.push(RiscvAir::ProgramMemory(program_memory_init));
         let byte = ByteChip::default();
         chips.push(RiscvAir::ByteLookup(byte));
+        let blake_2s_round = Blake2sRoundChip::new();
+        chips.push(RiscvAir::Blake2sRound(blake_2s_round));
```
Comment (wwared) [inline core/src/stark/air.rs:188]: This should be moved up to after `bls12381_g1_decompress` above I believe

**Comment by adr1anh** [comment]
The constraints look good, though I'm taking a better look to see if we can do some optimizations (for example, by using transition constraints and chaining the calls to the compression function, avoiding extra memory accesses). 

I was wondering where the test vectors came from. There are some proposed in the [RFC](https://www.rfc-editor.org/rfc/rfc7693.txt) and was wondering if they cam from there. Would it also be possible to show a full end-to-end test that computes a full hash using the precompile. The classic would be checking `BLAKE2s-256("abc") = 50 8C 5E 8C 32 7C 14 E2 E1 A7 2B A3 4E EB 45 2F 37 45 8B 20 9E D6 3A 29 4D 99 9B 4C 86 67 59 82`.

**Comment by storojs72** [comment]
@adr1anh, this precompile implements single round function of Blake2s - not a whole compress function, so I extracted the test values by injecting printlns directly into the macro [code](https://github.com/argumentcomputer/RustCrypto-hashes/blob/zkvm/blake2/src/macros.rs#L144), using `Blake2s("")` initially. Then, in order to guarantee that it will work, I used sphinx program from https://github.com/argumentcomputer/zk-light-clients/pull/241 that runs single hashing using randomised input passed from the [following](https://github.com/argumentcomputer/zk-light-clients/blob/blake2s-program/kadena/light-client/src/lib.rs#L20) unit-test in Kadena LC, which compares the outputs of hashing performed in-circuit and out-of-circuit.

**Review by wwared — APPROVED** [review]

**Pull request merged**